### ISSUE TITLE
Socket leak in DockerD

### ISSUE THREAD

**Issue by msabansal** [moby/moby#35692]
**Description**
In our density runs we are seeing socket leak with latest dockerd. This only happens when the API communication happens over http. I suspect it's something related to hijacking where we fail to close the socket. I have tried debugging this but haven't made much headway because it happens in high density and i am not sure how to really track go's stack allocation. Any suggestions would be helpful. 

**Steps to reproduce the issue:**
1. Configure Hyperv and container in a VM
2.  Attach  kernel debugger to the VM
3.  Set a break point for every allocation or dellocation (ed nt!poolhittag 'EcpT' or ed nt!poolhittag 'EelA')
4.  Inside VM run  command: docker run --rm windowsservercore ipconfig

**Describe the results you received:**
Since allocation de-allocation is very high adding only the poolmon related details.

AFD tags are:
afd.sys - Afd context buffer 	AfdX	Paged	1394935	1284612	110323	25593504	231	6.022416
afd.sys - Afd endpoint structure 	AfdE	Nonp	636447	526107	110340	51207184	464	5.998821

0: kd> !afdkd.conn -b

In our case above command clearly showing socket leak in dockerd process(PID = 063C or 1596)

Connection            Stat Flags   SndB-cnt RcvB-cnt     Pid    Endpoint                 Remote Address...
ffffc18b08cc3590 Con  DRS   t 00000-00 00000-00 063c ffffc18b0a1276b0 [fe80::157d:6913:6273:bce2%9]:58699
ffffc18b10551480 Con  DRS   t 00000-00 00000-00 063c ffffc18b0a5d2e40 [fe80::157d:6913:6273:bce2%9]:58692
ffffc18b086c4240 Con  DRS   t 00000-00 00000-00 063c ffffc18b08248e40 [fe80::157d:6913:6273:bce2%9]:58685
ffffc18b08c29a70 Con  DRS   t 00000-00 00000-00 063c ffffc18b0a7cae40 [fe80::157d:6913:6273:bce2%9]:58678
ffffc18b12040aa0 Con  DRS   t 00000-00 00000-00 063c ffffc18b120d5e40 [fe80::157d:6913:6273:bce2%9]:58671
ffffc18b092cb010 Con  DRS   t 00000-00 00000-00 063c ffffc18b0ba45af0 [fe80::157d:6913:6273:bce2%9]:58664
ffffc18b098d7830 Con  DRS   t 00000-00 00000-00 063c ffffc18b0e876e40 [fe80::157d:6913:6273:bce2%9]:58657
ffffc18b06761190 Con  DRS   t 00000-00 00000-00 063c ffffc18b091ca5f0 [fe80::157d:6913:6273:bce2%9]:58650
ffffc18b0841cf00 Con  DRS   t 00000-00 00000-00 063c ffffc18b08651490 [fe80::157d:6913:6273:bce2%9]:58643
ffffc18b09078a40 Con  DRS   t 00000-00 00000-00 063c ffffc18b119b0e40 [fe80::157d:6913:6273:bce2%9]:58636
ffffc18b08963980 Con  DRS   t 00000-00 00000-00 063c ffffc18b0b72fe40 [fe80::157d:6913:6273:bce2%9]:58629

063c is the docker process

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
(paste your output here)
```

**Output of `docker info`:**

```
(paste your output here)
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
ping @crosbymichael @mlaventure PTAL

**Comment by thaJeztah** [comment]
Doh! I read `DockerD` as `ContainerD`

**Comment by johnstep** [comment]
I will look into this. Thanks.

**Comment by thaJeztah** [comment]
ping @johnstep did you have time to look into this?

**Comment by johnstep** [comment]
Yes, I spent significant time debugging this a while back, and confirmed handle leaks (and similar leaks do not exist on Linux). I have been looking into other things, and still have not determined the root cause, but will provide an update later this week.
